File: python/n404_LeftLeaf.py
# https://leetcode-cn.com/problems/sum-of-left-leaves/

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def sumOfLeftLeaves(self, root: TreeNode) -> int:
        if root is None: return 0
        if root.left is None:
            return self.sumOfLeftLeaves(root.right)
        if self.isLeaf(root.left):
            return root.left.val + self.sumOfLeftLeaves(root.right)
        return self.sumOfLeftLeaves(root.left) + self.sumOfLeftLeaves(root.right)

    def isLeaf(self, node: TreeNode) -> bool:
        if node.left is None and node.right is None:
            return True
        else:
            return False


if __name__ == "__main__":
    pass

File: cpp/n404_LeftLeaf.cpp
// https://leetcode-cn.com/problems/sum-of-left-leaves/
#include <vector>
#include <string>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int sumOfLeftLeaves(TreeNode* root) {
        if (root == NULL) return 0;
        if (root -> left == NULL) return sumOfLeftLeaves(root -> right);
        if (isLeaf(root -> left)) return (root -> left -> val + sumOfLeftLeaves(root -> right));
        return (sumOfLeftLeaves(root -> left) + sumOfLeftLeaves(root -> right));
    }

private:    
    bool isLeaf(TreeNode* node) {
        if (node -> left == NULL && node -> right == NULL)
            return true;
        else
            return false;
    }
};

int main(){
    return 0;
}